lib/internal/test_runner/test.js

```diff
@@ -29,13 +31,31 @@ const kMultipleCallbackInvocations = 'multipleCallbackInvocations';
 const kParentAlreadyFinished = 'parentAlreadyFinished';
 const kSubtestsFailed = 'subtestsFailed';
 const kTestCodeFailure = 'testCodeFailure';
+const kTestTimeoutFailure = 'testTimeoutFailure';
 const kDefaultIndent = '    ';
+const kDefaultTimeout = null;
 const noop = FunctionPrototype;
 const isTestRunner = getOptionValue('--test');
 const testOnlyFlag = !isTestRunner && getOptionValue('--test-only');
 // TODO(cjihrig): Use uv_available_parallelism() once it lands.
 const rootConcurrency = isTestRunner ? cpus().length : 1;
 
+
+function testTimeout(promise, timeout) {
+  if (timeout === kDefaultTimeout) {
+    return promise;
+  }
+  return PromiseRace([
```

Comment:
Not really a fan at all of using `Promise.race()` here in that it does not actually cancel the `promise` once it times out. I would much prefer a mechanism that builds properly on a test supporting `AbortSignal` to detect and properly handle cancelations. For instance, if I want a test that is canceled after 30 milliseconds, I would do something like...

```
test('it times out', { signal: AbortSignal.timeout(30) }, (t) => {
  t.signal.addEventListener('abort', () => {
    // logic to actually cancel the test.
  });
});
```

Reply:
Yeah see https://github.com/nodejs/node/pull/43554
I think we should support both